README.md
# ember-burger-menu

[![CI](https://github.com/adopted-ember-addons/ember-burger-menu/actions/workflows/ci.yml/badge.svg)](https://github.com/adopted-ember-addons/ember-burger-menu/actions/workflows/ci.yml)
[![npm version](https://badge.fury.io/js/ember-burger-menu.svg)](http://badge.fury.io/js/ember-burger-menu)

An off-canvas sidebar component with a collection of animations and styles using CSS transitions

## Features

- Easy to use & setup off-canvas menu
- Mix and match from many menu & menu item animations
- Swipe gesture support with changeable thresholds
- Easily create your own animations

## Compatibility

* Ember.js v3.24 or above
* Ember CLI v3.24 or above
* Node.js v12 or above

## Installation

```
ember install ember-burger-menu
```

### Sass

Installing ember-burger-menu should also install ember-cli-sass and automatically create a scss file under `app/styles/app.scss` with

```sass
// app/styles/app.scss

@import 'ember-burger-menu';
```

#### Overriding Variables

Using sass, you can override default variables and easily change the default behavior of ember-burger-menu.
See [variables.scss](app/styles/ember-burger-menu/variables.scss) for a list of variables you can change.

```sass
// app/styles/app.scss

// Burger Menu Overrides
$bm-transition-duration: 0.3s;
$bm-overlay-background: rgba(0, 0, 0, 0.7);

// Import all the styles!
@import 'ember-burger-menu';
```

#### Import Only What You Need

Using sass, you can import only the styles you need for the animations you use.

```sass
// Core Styles
@import 'ember-burger-menu/variables';
@import 'ember-burger-menu/structure';

// Animations
@import 'ember-burger-menu/animations/push';
@import 'ember-burger-menu/animations/menu-item/stack';
```

## Helpful Links

- ### [Live Demo](http://adopted-ember-addons.github.io/ember-burger-menu)

- ### [Changelog](CHANGELOG.md)

## Looking for help?
If it is a bug [please open an issue on GitHub](http://github.com/adopted-ember-addons/ember-burger-menu/issues).

## Animations

### Menu Animations

- slide
- reveal
- push
- fall-down
- open-door
- push-rotate
- rotate-out
- scale-up
- scale-down
- scale-rotate
- slide-reverse
- squeeze

### Menu Item Animations

- push
- stack

## Usage

This addon utilizes contextual components to be able to correctly control and animate necessary elements.

```hbs
{{#burger-menu as |burger|}}
  {{#burger.menu itemTagName="li" as |menu|}}
    <button {{on "click" burger.state.closeMenu}}>Close</button>

    <ul>
      {{#menu.item}}
        <LinkTo @route="features">
          Features
        </LinkTo>
      {{/menu.item}}

      {{#menu.item}}
        <LinkTo @route="about">
          About
        </LinkTo>
      {{/menu.item}}

      {{#menu.item}}
        <LinkTo @route="contact">
          Contact Us
        </LinkTo>
      {{/menu.item}}
    </ul>
  {{/burger.menu}}

  {{#burger.outlet}}
    <button {{on "click" burger.state.toggleMenu}}>Menu</button>
    {{outlet}}
  {{/burger.outlet}}
{{/burger-menu}}
```

## Components

### `{{burger-menu}}`

#### Options

- #### `open`

  The current open state of the menu.

  **Default: false**

- #### `animation`

  The menu animation. See [Animations](#menu-animations) for the list of available animations.

  **Default: slide**

- #### `itemAnimation`

  The menu item animation. See [Item Animations](#menu-item-animations) for the list of available item animations.

  **Default: null**

- #### `position`

  The menu's open position. Can either be _left_ or _right_

  **Default: left**

- #### `width`

  The menu's width (in px).

  **Default: 300**

- #### `locked`

  Lock the menu in its current open / closed state.

  **Default: false**

- #### `customAnimation`

  Override of the menu's styles with your own implementation. See [Custom Animations](#custom-animations) for more details.

- #### `translucentOverlay`

  Whether the outlet has a translucent overlay over it once the menu is opened.

  **Default: true**

- #### `dismissOnClick`

  Whether the menu can be dismissed when clicking outside of it.

  **Default: true**

- #### `dismissOnEsc`

  Whether the menu can be closed when pressing the ESC key.

  **Default: true**

- #### `gesturesEnabled`

  Whether the menu can be open / closed using gestures. The only available gesture currently is swipe.

  **Default: true**

- #### `minSwipeDistance`

  The minimum distance (in px) the user must swipe to register as valid.

  **Default: 150**


- #### `maxSwipeTime`

  The maximum amount of time (in ms) it must take the user to swipe to be registered as valid .

  **Default: 300**

### `{{burger.outlet}}`

This component is where all your content should be placed.

### `{{burger.menu}}`

Everything rendered here will be inside the menu.

#### Options

- #### `itemTagName`

  The default tagName that will be used by the `{{menu.item}}` component.

  **Default: div**

- #### `dismissOnItemClick`

  Close the menu on click of a `{{menu.item}}`.

  **Default: false**

#### Actions

- #### `onOpen`

  Triggered when the menu is opening

- #### `onClose`

  Triggered when the menu is closing

### `{{menu.item}}`

The individual menu item. This is required if you have specified an [itemAnimation](#itemanimation).

#### Options

- #### `dismissOnClick`

  Close the menu on click.

  **Default: false**

## The Menu State

The `{{burger-menu}}` component exposes multiple contextual components, but it also exposes a state object.

You can use the menu state object to modify pretty much any property.

- `open`
- `width`
- `position`
- `animation`
- `itemAnimation`
- `customAnimation`

The state object also exposes some actions:

- #### `open`

  ```hbs
  <button {{on "click" burger.state.openMenu}}>Open</button>
  ```

- #### `close`

  ```hbs
  <button {{on "click" burger.state.closeMenu}}>Close</button>
  ```

- #### `toggle`

  ```hbs
  <button {{on "click" burger.state.toggleMenu}}>Toggle</button>
  ```

# Custom Animations

If you're not impressed with the in-house animations and want to create your own, all you have to do is pass the following class to the `customAnimation` property in the `{{burger-menu}}` component. If you think your animation would be a good addition to the existing collection, feel free to open a PR with it!

```js
import Animation from 'ember-burger-menu/animations/base';

export default Animation.extend({
  // CSS class names to be able to target our menu
  animation: 'my-custom-animation',
  itemAnimation: 'my-custom-item-animation',

  container(open, width, right) {
    return {};
  },

  outlet(open, width, right) {
    return {
      transform: open ? right ? `translate3d(-${width}px, 0, 0)` : `translate3d(${width}px, 0, 0)` : ''
    };
  },

  menu(open, width, right) {
    return {};
  },

  menuItem(open, width, right, index) {
    return {
      transform: open ? '' : `translate3d(0, ${(index + 1) * 500}px, 0)`
    };
  }
});

```

_**Note:** You don't need to worry about prefixing your CSS attributes as it will be done for you._

If you need to add some base CSS to your animation, you can target the menu as such:

```sass
.ember-burger-menu.bm--my-custom-animation {
  #{$bm-menu} {}
  > .bm-outlet {}

  &.is-open {
    #{$bm-menu} {}
    > .bm-outlet {}
  }
}
```

And the menu items as such:

```sass
.ember-burger-menu.bm-item--my-custom-item-animation {
  #{$bm-menu} {
    .bm-menu-item {}
  }

  &.is-open {
    #{$bm-menu} {
      .bm-menu-item {}
    }
  }
}
```

To use our new custom animation, all we have to do is pass the class to
the `customAnimation` option in the `{{burger-menu}}` component.

```js
import MyCustomAnimation from 'path/to/my-custom-animation';

export default Ember.Component.extend({
  MyCustomAnimation
});
```

```hbs
{{#burger-menu customAnimation=MyCustomAnimation}}
  ...
{{/burger-menu}}
```

# Contributing
See the [Contributing](CONTRIBUTING.md) guide for details.


CHANGELOG.md
# Changelog


## v4.0.0 (2022-06-03)

#### :boom: Breaking Change
* [#152](https://github.com/adopted-ember-addons/ember-burger-menu/pull/152) ember-auto-import 2, node >= 12, update all the things ([@rwwagner90](https://github.com/rwwagner90))
* [#124](https://github.com/adopted-ember-addons/ember-burger-menu/pull/124) Update ember and various things ([@ttill](https://github.com/ttill))

#### Committers: 2
- Robert Wagner ([@rwwagner90](https://github.com/rwwagner90))
- Till Theato ([@ttill](https://github.com/ttill))

## v3.3.3

### Pull Requests

- [#118](https://github.com/adopted-ember-addons/ember-burger-menu/pull/118) resolve a later version of `matches-selector` _by [@noslouch](https://github.com/noslouch)_

## v3.3.2

### Pull Requests

- [#116](https://github.com/adopted-ember-addons/ember-burger-menu/pull/116) fix(element-closest): IE11 fix to use implementation via npm package `element-closest` _by [@alexdiliberto](https://github.com/alexdiliberto)_

## v3.3.1

### Pull Requests

- [#111](https://github.com/adopted-ember-addons/ember-burger-menu/pull/111) Completely remove usage of jQuery _by [@josemarluedke](https://github.com/josemarluedke)_
- [#112](https://github.com/adopted-ember-addons/ember-burger-menu/pull/112) Remove unused ember-data dependency _by [@josemarluedke](https://github.com/josemarluedke)_
- [#113](https://github.com/adopted-ember-addons/ember-burger-menu/pull/113) Only include used bootstrap styles _by [@josemarluedke](https://github.com/josemarluedke)_

## v3.3.0

### Pull Requests

- [#110](https://github.com/adopted-ember-addons/ember-burger-menu/pull/110) Upgrade to latest Ember CLI and update dependencies _by [@josemarluedke](https://github.com/josemarluedke)_

## v3.2.0

### Pull Requests

- [#109](https://github.com/adopted-ember-addons/ember-burger-menu/pull/109) Remove usage of jquery in components _by [@josemarluedke](https://github.com/josemarluedke)_

## v3.1.0

### Pull Requests

- [#100](https://github.com/adopted-ember-addons/ember-burger-menu/pull/100) Upgrade to Ember 2.18 _by [@mixonic](https://github.com/mixonic)_
- [#104](https://github.com/adopted-ember-addons/ember-burger-menu/pull/104) Update all the things

## v3.0.2

### Pull Requests

- [#102](https://github.com/adopted-ember-addons/ember-burger-menu/pull/102) Update ember-cli-dependency-checker

## v3.0.1

### Pull Requests

- [#98](https://github.com/adopted-ember-addons/ember-burger-menu/pull/98) Fix for lifeline selector issue _by [@donaldwasserman](https://github.com/donaldwasserman)_

## v3.0.0

### Pull Requests

- [#88](https://github.com/adopted-ember-addons/ember-burger-menu/pull/88) Better Sass _by [@offirgolan](https://github.com/offirgolan)_

### Release Notes

- Hard dependency on sass (now required in the host application)
  - Allows to override various variables
  - Allows to selective import only the sass files you need
- More fine grain selector matching to support multi-level nested burger menus
- **[BREAKING]** `slide-shrink` renamed to `squeeze`
- **[BREAKING]** Menu item class has now been moved to the top level `ember-burger-menu` element. It can now be accessed via `.ember-burger-menu.bm-item--stack`
- **[BREAKING]** Please see [Custom Animations](https://github.com/adopted-ember-addons/ember-burger-menu#custom-animations) for the new sass structure for custom animations

## v2.0.1

### Pull Requests

- [#80](https://github.com/adopted-ember-addons/ember-burger-menu/pull/80) Fastboot _by [@mayko780](https://github.com/mayko780)_

## v2.0.0

### Pull Requests

- [#79](https://github.com/adopted-ember-addons/ember-burger-menu/pull/79) [FEATURE] Multiple Menu Support _by [@offirgolan](https://github.com/offirgolan)_

### Release Notes

- Multiple menus can now co-exist in your application
- Improved event listeners
- **[BREAKING]** Removed `burger-menu` service as it was used to back a single menu state and no longer makes sense to have.
- **[BREAKING]** When the menu is in a `locked` state, actions can no longer change the state of the menu.
- **[BREAKING]** The default height is no longer set to `100vh`, instead it is set to `100%`. If you upgrade and your menu is no longer visible, please check the available viewport height.

## v1.1.3

### Pull Requests

- [#69](https://github.com/adopted-ember-addons/ember-burger-menu/pull/69) [BUGFIX] Remove preventDefault from passive event listener touchMove _by [@offirgolan](https://github.com/offirgolan)_

## v1.1.2

### Pull Requests

- [#62](https://github.com/adopted-ember-addons/ember-burger-menu/pull/62) [FEATURE] Slide Shrink Animation _by [@offirgolan](https://github.com/offirgolan)_

## v1.1.1

### Pull Requests

- [#42](https://github.com/adopted-ember-addons/ember-burger-menu/pull/42) Update ember-cli-sass to the latest version 🚀 _by [@offirgolan/greenkeeper](https://github.com/offirgolan/greenkeeper)_

## v1.1.0

### Pull Requests

- [#29](https://github.com/adopted-ember-addons/ember-burger-menu/pull/29) [FEATURE] Locked + gesturesEnabled + use ember-lifeline _by [@offirgolan](https://github.com/offirgolan)_

## v1.0.2

### Pull Requests

- [#21](https://github.com/adopted-ember-addons/ember-burger-menu/pull/21) Make height 100vh _by [@rwwagner90](https://github.com/rwwagner90)_

## v1.0.1

### Pull Requests

- [#13](https://github.com/adopted-ember-addons/ember-burger-menu/pull/13) [BUGFIX] Use originalEvent to access touches _by [@offirgolan](https://github.com/offirgolan)_

## v1.0.0

### Pull Requests

- [#10](https://github.com/adopted-ember-addons/ember-burger-menu/pull/10) [FEATURE] Dismiss menu on item click _by [@offirgolan](https://github.com/offirgolan)_
- [#11](https://github.com/adopted-ember-addons/ember-burger-menu/pull/11) [FEATURE] Use a service to hold the menu state _by [@offirgolan](https://github.com/offirgolan)_

## v0.2.1

### Pull Requests


addon/-private/state.js
/* eslint-disable ember/no-computed-properties-in-native-classes */
import EmberObject, { action, computed } from '@ember/object';
import getAnimation from 'ember-burger-menu/animations';
import { tracked } from '@glimmer/tracking';

export default class State extends EmberObject {
  @tracked open = false;
  @tracked locked = false;
  @tracked width = 300;
  @tracked position = 'left';
  @tracked animation = 'slide';

  @tracked minSwipeDistance = 150;
  @tracked maxSwipeTime = 300;

  @tracked itemAnimation = null;
  @tracked customAnimation = null;

  @computed('animation', 'itemAnimation', 'customAnimation')
  get styles() {
    let animation = this.animation;
    let itemAnimation = this.itemAnimation;
    let customAnimation = this.customAnimation;

    return getAnimation(customAnimation || animation, itemAnimation).create();
  }

  @action
  openMenu(e) {
    e?.stopPropagation?.();
    if (!this.locked) {
      this.open = true;
    }
  }

  @action
  closeMenu(e) {
    e?.stopPropagation?.();
    if (!this.locked) {
      this.open = false;
    }
  }

  @action
  toggleMenu(e) {
    e?.stopPropagation?.();
    if (!this.locked) {
      this.open = !this.open;
    }
  }
}
